### PR TITLE
Expand ~ in paths for cmdline_for_next_invocation

### PR DESCRIPTION
**Pull request by antonzhukovin**
This PR implements a solution for #234 and replaces `~` when found in `location` or `path` settings for benchmark suites and executors respectively.

### Major Change

To make this work, we now require Python 3.8 as minimum version.

### Limitations

We use `shlex.split()` for working with the commands, which can change the type of quote used for strings in commands, and may change how shells interpolate variables.
Might need to be revisited if that causes issues.

### CODE REVIEW

**Comment by smarr** [comment]
I added an initial description for the PR, just so that we keep track of the context, and what exactly is done.

Could you please remove the `build` folder that got added? It should not be part of the repository.

**Comment by antonzhukovin** [comment]
Removed build folder from repository

File: rebench/denoise.py [hunk 1]
```diff
@@ -19,39 +19,7 @@
     rebench_version = "unknown"
 
 
-class CommandsPaths:
```
Comment (smarr) [inline rebench/denoise.py:22]: Not sure what's going on here, but this looks like you're reverting a change that was merged yesterday.

File: rebench/denoise.py [hunk 2]
```diff
@@ -430,8 +402,6 @@ def main_func():
     arg_parser = _shell_options()
     args, remaining_args = arg_parser.parse_known_args()
```
Comment (smarr) [inline rebench/denoise.py:404]: Looks like all changes in the denoise.py file revert changes.

File: rebench/executor.py [hunk 3]
```diff
@@ -28,7 +28,6 @@
 from time import time
 
 from . import subprocess_with_timeout as subprocess_timeout
-from .denoise import paths as denoise_paths
```
Comment (smarr) [inline rebench/executor.py:31]: Reverts recently merged change.

File: rebench/executor.py [hunk 4]
```diff
@@ -343,14 +342,13 @@ def _construct_cmdline(self, run_id, gauge_adapter):
             cmdline += "sudo "
             if run_id.env:
                 cmdline += "--preserve-env=" + ','.join(run_id.env.keys()) + " "
-            cmdline += denoise_paths.get_denoise() + " "
```
Comment (smarr) [inline rebench/executor.py:346]: Reverts recently merged change.

File: rebench/executor.py [hunk 5]
```diff
@@ -343,14 +342,13 @@ def _construct_cmdline(self, run_id, gauge_adapter):
...
+            cmdline += "rebench-denoise "
             if not self._use_nice:
                 cmdline += "--without-nice "
             if not self._use_shielding:
                 cmdline += "--without-shielding "
             if run_id.is_profiling():
                 cmdline += "--for-profiling "
-            cmdline += "--cset-path " + denoise_paths.get_cset() + " "
```
Comment (smarr) [inline rebench/executor.py:353]: Reverts recently merged change.

File: rebench/model/run_id.py [hunk 6]
```diff
@@ -278,7 +279,11 @@ def cmdline(self):
 
     def cmdline_for_next_invocation(self):
         """Replace the invocation number in the command line"""
-        return self.cmdline() % {'invocation': self.completed_invocations + 1}
+        cmdline = self.cmdline() % {'invocation': self.completed_invocations + 1}
+        print(cmdline)
+        print(os.path.expanduser(cmdline))
+        return os.path.expanduser(cmdline)
+        # return self.cmdline() % {'invocation': self.completed_invocations + 1}
```
Comment (smarr) [inline rebench/model/run_id.py:286]:
```suggestion
        return os.path.expanduser(cmdline)
```

Please remove the prints and the commented out code.

File: rebench/tests/features/issue_42_test.py [hunk 7]
```diff
@@ -136,17 +136,17 @@ def test_construct_cmdline_env_vars_set_as_expected(self):
         runs = list(cnf.get_runs())
         ex = Executor(runs, True, self.ui, use_nice=True, use_shielding=True)
 
-        self.assertRegex(ex._construct_cmdline(runs[0], TimeManualAdapter(False, ex)),
-                         r" --preserve-env=IMPORTANT_ENV_VARIABLE,ALSO_IMPORTANT [^\s]*rebench-denoise")
+        self.assertIn(" --preserve-env=IMPORTANT_ENV_VARIABLE,ALSO_IMPORTANT rebench-denoise",
```
Comment (smarr) [inline rebench/tests/features/issue_42_test.py:139]: Changes in this file revert recently merged changes.

File: rebench/tests/features/vm_58a.sh [hunk 8]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (smarr) [inline rebench/tests/features/vm_58a.sh:1]: Not sure where these files come from.

File: rebench/tests/features/vm_58b.sh [hunk 9]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (smarr) [inline rebench/tests/features/vm_58b.sh:1]: Not sure where this file comes from.

File: rebench/tests/payload.json [hunk 10]
```diff
@@ -0,0 +1 @@
...,"c":3}]},{"in":7,"it":1,"m":[{"v":104.97925087243438,"c":0},{"v":417.1577642030033,"c":1},{"v":226.5407673638457,"c":2},{"v":790.8086796292213,"c":3}]},{"in":8,"it":1,"m":[{"v":108.37340203187172,"c":0},{"v":406.235737911574,"c":1},{"v":251.62574814046243,"c":2},{"v":793.5576172779321,"c":3}]},{"in":9,"it":1,"m":[{"v":102.70641322763063,"c":0},{"v":428.4745165545249,"c":1},{"v":227.10261093889022,"c":2},{"v":754.5530406378048,"c":3}]},{"in":10,"it":1,"m":[{"v":105.37646892876835,"c":0},{"v":421.20073385317255,"c":1},{"v":239.45002723734973,"c":2},{"v":744.86416978315,"c":3}]}]}],"criteria":[{"c":"part1","u":"ms","i":0},{"c":"part2","u":"ms","i":1},{"c":"part3","u":"ms","i":2},{"c":"total","u":"ms","i":3}],"env":{"hostName":"test","userName":"test"},"startTime":"2024-07-03T14:45:45.564985+00:00","source":{"repoURL":"http://repo.git","branchOrTag":"master","commitId":"1de0e22fb3e9578b3aeb7a994481ead300fa977f","commitMsg":"Make the output for the current benchmark that's running more compact (#237)\n\n","authorName":"Stefan Marr","committerName":"GitHub","authorEmail":"git@stefan-marr.de","committerEmail":"noreply@github.com"},"projectName":"Persistency Test","experimentName":"Test"}
```
Comment (smarr) [inline rebench/tests/payload.json:1]: Please remove the .json files. They also should not be added.

**Review by smarr — COMMENTED** [review]
Hm, there are various changes that revert recently merged stuff. And a few other odd things, like added files that shouldn't be added. Would be good to remove/undo those.
Happy to help.

**Comment by antonzhukovin** [comment]
Removed debug prints and commented codes

**Comment by smarr** [comment]
I rebased the PR on top of the latest master branch.

I think, I'd also still like another test.
We may want to support `~` in the command of a benchmark.
E.g. here: https://github.com/smarr/ReBench/pull/240/files#diff-929519fa87febd9e4d8db9c1e9b9b3b05721f4c5360363cbac763ea5d89de7f8R16

Perhaps the command would be `command: 'TestBenchMarks ~/suiteFolder/%(benchmark)s'` and should be expanded properly.

**Comment by smarr** [comment]
Thank you!

**Pull request merged**